src/libutil/filesystem.cpp

```diff
@@ -1013,4 +1013,129 @@ Filesystem::scan_for_matching_filenames(const std::string &pattern_,
     return true;
 }
 
+
+
+Filesystem::IOFile::IOFile (string_view filename, Mode mode)
+    : IOProxy(filename,mode)
+{
+    // Call Filesystem::fopen since it handles UTF-8 file paths on Windows,
+    // which std fopen does not.
+    m_file = Filesystem::fopen (m_filename.c_str(), mode == Write ? "wb" : "rb");
+    if (! m_file)
+        mode = Closed;
+    m_auto_close = true;
+    if (mode == Read)
+        m_size = Filesystem::file_size (filename);
+}
+
+Filesystem::IOFile::IOFile (FILE *file, Mode mode)
+    : IOProxy("",mode), m_file(file)
+{
+    if (mode == Read) {
+        m_pos = ftell (m_file);             // save old position
+        fseek (m_file, 0, SEEK_END);        // seek to end
+        m_size = size_t(ftell (m_file));    // size is end position
+        fseek (m_file, m_pos, SEEK_SET);    // restore old position
+    }
+}
+
+Filesystem::IOFile::~IOFile ()
+{
+    if (m_auto_close)
+        close();
+}
+
+void
+Filesystem::IOFile::close ()
```

Comment:
Is there any case where we would try to re-read the file after it has been closed? I thought a few plugins do this.

Reply:
The plugins that I have added support to do not need to close until they are truly done. If later we tackle one that makes a habit of opening and closing, we'll either need to change the plugin so that it doesn't do that (will a seek to the beginning do?), or else maybe we'll need to add a way to do an explicit open() to IOFile, which so far I've avoided.

The trouble with explicit IOFile::open() is that in the form that simply wraps an already-open FILE* passed by the caller, the IOFile doesn't know what the filename is, in order to reopen it if it's explicitly closed. So that would also throw a bit of a wrench into the API, as well as pop up new problem cases where you maybe couldn't construct an IOFile from parts of the code that had the FILE* passed to it but didn't know the filename.

I admit to being a little unsure how to proceed with this, but I'm inclined to push forward until we have identified an actual use case where it really does need to fully close and re-open.

Reply:
I'll also point out that if you are using an ImageBuf, you do the same thing -- pass a config spec, which contains the pointer to the ioproxy. And it's even ok if that ImageBuf is "ImageCache-backed", because the buf will communicate the config to the underlying IC!

Reply:
That all makes sense. Cool!

Reply:
This is only in master for now, semi-experimental. I'm assuming that we'll probably diddle with the semantics a bit in the coming months as people try to use it and figure out what I missed in the design.

One more aside: On Unix/Linux, it would be reasonable to have IOFile::close() also do nothing and then be totally safe in this way. The only reason we need to support explicit true close (prior to destroying the IOFile) is because on Windows, an open handle locks the file in a way that other processes can't open it (if either or the other process have opened it for writing).

Reply:
I get the Windows constraint, but as you point out, if you have an actual, standalone image file, you don't need to use an IOProxy, and if the proxy does or will ever need to support reopening/further-access after an explicit close(), then any bespoke memory or mmapped-based implementation will need to keep the resource alive, unless it can know how to reconstruct it.  So it doesn't seem like there's a likely practical case for close() being useful (assuming you use the nonproxy API for actual files), to warrant the extra complexity?

Reply:
There are certain situations where you'd use an IOProxy **or** rely on the default behavior for an ordinary file. You wouldn't expect a calling app to pass in an IOFile as the proxy, there's no point. But note that this patch was straightforward for both OpenEXR and PNG only because they use 3rd party libraries that happen to already have a way to have a 'write' callback, so only the callback setup had to change, to be one that uses the proxy.

But many of the formats we support don't have their own libraries with I/O overrides, and we do all our own file handling. As an example, I wouldn't add proxy support to pnminput.cpp by doubling the size of the source and having every method cloned separately handle to handle the FILE* case and the IOProxy case. A simpler solution is to rewrite the whole thing to be IOProxy throughout, and then it's a simple matter of whether it's a passed-in proxy from the caller (probably a memreader), or an internally-constructed IOFile proxy for the default case of going to the file.

So now we're back to a case where you want a close() from the calling application to do a real close right then, so it's not holding a file handle open (taking up the resource on Linux/OSX, and also locking, ick, on Windows).